src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java

```diff
@@ -1,31 +1,272 @@
+                    }
+
+                    applyOverrides(upperBounds, Collections.emptyMap(), shadow, getLog());
+                }
+            }
+
+            /*
+             * At this point, the model has been updated as the user has requested. We now redo
+             * resolution and compare the new resolution to the original in order to account for
+             * updates to transitive dependencies that are not present in the model. Anything that
+             * was updated in the new resolution needs to be updated in the test classpath. Anything
+             * that was removed in the new resolution needs to be removed from the test classpath.
+             * Anything that was added in the new resolution needs to be added to the test
+             * classpath.
+             */
+            Set<Artifact> resolved = resolveDependencies(shadow);
+            effectiveArtifacts = wrap(new Artifacts(resolved));
+            Map<String, String> newResolution = new HashMap<>();
+            for (Artifact artifact : resolved) {
+                newResolution.put(toKey(artifact), artifact.getVersion());
+            }
+            for (Map.Entry<String, String> entry : newResolution.entrySet()) {
+                if (originalResolution.containsKey(entry.getKey())) {
+                    // Present in both old and new resolution: check for update.
+                    String originalVersion = originalResolution.get(entry.getKey());
+                    String newVersion = entry.getValue();
+                    if (!newVersion.equals(originalVersion)) {
+                        updates.put(entry.getKey(), newVersion);
+                    }
+                } else {
+                    // Present in new resolution but not old: addition.
+                    additions.put(entry.getKey(), entry.getValue());
+                }
+            }
+            for (Map.Entry<String, String> entry : originalResolution.entrySet()) {
+                if (!newResolution.containsKey(entry.getKey())) {
+                    // Present in old resolution but not new: deletion.
+                    deletions.put(entry.getKey(), entry.getValue());
+                }
+            }
+            getLog().info("New dependency tree:");
+            MavenSession shadowSession = session.clone();
+            shadowSession.setCurrentProject(shadow);
+            shadow.setArtifacts(resolved);
+            MojoExecutor.executeMojo(
+                    MojoExecutor.plugin(
+                            MojoExecutor.groupId("org.apache.maven.plugins"),
+                            MojoExecutor.artifactId("maven-dependency-plugin")),
+                    MojoExecutor.goal("tree"),
+                    MojoExecutor.configuration(
+                            MojoExecutor.element(MojoExecutor.name("scope"), Artifact.SCOPE_TEST)),
+                    MojoExecutor.executionEnvironment(shadow, shadowSession, pluginManager));
+            if (getLog().isDebugEnabled()) {
```

Comment:
NIT: I would expect this to be at info level and the dependency:tree to be debug.

Reply:
For what reason do you have this expectation?

Reply:
I think the argument about it being too verbose makes sense. I swapped them in commit 2ca309b.